graylog2-server/src/main/java/org/graylog2/security/realm/LdapUserAuthenticator.java

```diff
@@ -159,12 +141,83 @@ protected AuthenticationInfo doGetAuthenticationInfo(AuthenticationToken authtok
         return null;
     }
 
+    protected LdapNetworkConnection openLdapConnection(LdapSettings ldapSettings) throws LdapException {
+        final LdapConnectionConfig config = new LdapConnectionConfig();
+        config.setLdapHost(ldapSettings.getUri().getHost());
+        config.setLdapPort(ldapSettings.getUri().getPort());
+        config.setUseSsl(ldapSettings.getUri().getScheme().startsWith("ldaps"));
+        config.setUseTls(ldapSettings.isUseStartTls());
+        if (ldapSettings.isTrustAllCertificates()) {
+            config.setTrustManagers(new TrustAllX509TrustManager());
+        }
+        config.setName(ldapSettings.getSystemUserName());
+        config.setCredentials(ldapSettings.getSystemPassword());
+        return ldapConnector.connect(config);
+    }
+
+    protected LdapEntry searchLdapUser(LdapNetworkConnection connection, String principal, LdapSettings ldapSettings) throws LdapException, CursorException {
+        return ldapConnector.search(connection,
+                ldapSettings.getSearchBase(),
+                ldapSettings.getSearchPattern(),
+                ldapSettings.getDisplayNameAttribute(),
+                principal,
+                ldapSettings.isActiveDirectory(),
+                ldapSettings.getGroupSearchBase(),
+                ldapSettings.getGroupIdAttribute(),
+                ldapSettings.getGroupSearchPattern());
+    }
+
     public boolean isEnabled() {
         final LdapSettings ldapSettings = ldapSettingsService.load();
         return ldapSettings != null && ldapSettings.isEnabled();
     }
 
     @Nullable
+    public User syncLdapUser(String principal) {
+
+        final LdapSettings ldapSettings = ldapSettingsService.load();
+        if (ldapSettings == null || !ldapSettings.isEnabled()) {
+            LOG.trace("LDAP is disabled, skipping");
+            return null;
+        }
+
+        // do not try to look a token up in LDAP if there is no principal or password
+        if (isNullOrEmpty(principal)) {
+            LOG.debug("Principal wsa empty. Not trying to sync user with LDAP.");
```

Comment:
Not originally from your changes, but it would be a great opportunity to fix this typo ("wsa" ➡️ "was"). 😉